numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1315,8 +1315,10 @@ npyiter_check_casting(int nop, PyArrayObject **op,
         printf("\n");
 #endif
         /* If the types aren't equivalent, a cast is necessary */
-        if (op[iop] != NULL && !PyArray_EquivTypes(PyArray_DESCR(op[iop]),
-                                                     op_dtype[iop])) {
+        npy_intp view_offset = NPY_MIN_INTP;
+        if (op[iop] != NULL && !(PyArray_SafeCast(
+                    PyArray_DESCR(op[iop]), op_dtype[iop], &view_offset,
+                    NPY_NO_CASTING, 1) && view_offset == 0)) {
```

Comment:
I do wonder if we should have a clear way to spell it (up to changing the meaning of `PyArray_EquivTypes` to match this and changing the use where we want the meaning that allows `string_dtype1 == string_dtype2` where needed).